### ISSUE TITLE
p2sgrad, not converge

### ISSUE THREAD

**Issue by LuckyZxy182** [nii-yamagishilab/project-NN-Pytorch-scripts#15]
I try implement P2Sgrad using face recognition dataset MS1M v1c.
learning rate 0.1, not pretrained, not converge
learning rate 1e-3, pretrained, not converge

**Comment by TonyWangX** [comment]
Hi, sorry for hear that.

I can help if you provide more concrete information.

1. What do you mean by "not converge"?  ---- training loss does not decrease at all?  training loss does decrease but not for loss on development? loss goes to NaN?

2. "I try implement P2Sgrad" --- Do you mean that you tried to implement P2SGrad by yourself but did not get it work? Or did you tried the P2SGrad implementation in repository? 
For the former case, please share your code;
For the later case, please carefully check the data format when using the code in this repository. The DNNs in this repo assumes that the input is a speech data in shape (batch, length, feat_dim) (see more details in https://github.com/nii-yamagishilab/project-NN-Pytorch-scripts/blob/master/tutorials/c02_tensor_shape.ipynb). 
They may not be applicable to face images.

3. Have you tried other commonly used training criteria such as cross-entropy with plain softmax? Do they work? 
If they work, the issue is on P2SGrad;
if not, I would suggest checking the implementation of other parts of the model (as mentioned in comment 2 above).

**Comment by LuckyZxy182** [comment]
@TonyWangX Thanks for getting back to me. "not converge" means that  training loss goes to NaN. Based on your code, I construct P2Sgrad head for face recoginition. Unfortunately, it seems not work. Using same data and cosface margin+softmax-cross-entropy loss, it works well. Here is my code: 

`import math
import torch
import torch.nn as nn
import torch.nn.functional as F

class P2SHead(nn.Module):
    def __init__(self,
               in_channels,
               num_classes,
               drop_rate=0.):
        super(P2SHead, self).__init__()
        self.in_channels = in_channels
        self.num_classes = num_classes
        self.drop_rate = drop_rate
        self.m_loss = nn.MSELoss()
        self._init_layers()

    def _init_layers(self):
        self.weight = nn.Parameter(torch.Tensor(self.num_classes, self.in_channels))
        nn.init.kaiming_uniform_(self.weight, a=math.sqrt(5))

    def loss(self, logits, labels):
        losses = dict()
        with torch.no_grad():
            onehot_label = torch.zeros_like(logits)
            onehot_label.scatter_(1, labels.view(-1, 1), 1)
        loss_mse = self.m_loss(logits, onehot_label)

        losses['loss_mse'] = loss_mse
        return losses

    def forward(self, x, gt_label):
        # x:[batch_size, in_channels]
        if self.drop_rate:
            x = F.dropout(x, p=float(self.drop_rate), training=self.training)
        logits = F.linear(F.normalize(x), F.normalize(self.weight))
        logits = logits.clamp(-1, 1)
        gt_label = gt_label.long()
        losses = self.loss(logits, gt_label)
        return losses`

**Comment by LuckyZxy182** [comment]
@TonyWangX Thanks for getting back to me. "not converge" means that the loss goes to NaN. Based on your code, I construct my P2Sgrad head for face recognition task. The input in my forward function is in shape (batch_size, feat_dim). Using same data, same backbone and cosface margin + softmax-cross-entropy loss, the model works well. Here is my code.

```
import math
import torch
import torch.nn as nn
import torch.nn.functional as F

class P2SHead(nn.Module):
    def __init__(self,
               in_channels,
               num_classes,
               drop_rate=0.):
        super(P2SHead, self).__init__()
        self.in_channels = in_channels
        self.num_classes = num_classes
        self.drop_rate = drop_rate
        self.m_loss = nn.MSELoss()
        self._init_layers()

    def _init_layers(self):
        self.weight = nn.Parameter(torch.Tensor(self.num_classes, self.in_channels))
        nn.init.kaiming_uniform_(self.weight, a=math.sqrt(5))

    def loss(self, logits, labels):
        losses = dict()
        with torch.no_grad():
            onehot_label = torch.zeros_like(logits)
            onehot_label.scatter_(1, labels.view(-1, 1), 1)
        loss_mse = self.m_loss(logits, onehot_label)

        losses['loss_mse'] = loss_mse
        return losses

    def forward(self, x, gt_label):
        if self.drop_rate:
            x = F.dropout(x, p=float(self.drop_rate), training=self.training)
        logits = F.linear(F.normalize(x), F.normalize(self.weight))
        logits = logits.clamp(-1, 1)
        gt_label = gt_label.long()
        losses = self.loss(logits, gt_label)
        return losses
```

**Comment by TonyWangX** [comment]
Thanks for sharing the code. 

I plugged the code you shared to a speech anti-spoofing model and trained it on a speech dataset.
(Initial learning rate is 0.0003, no pre-training)

The performance is similar to what I got using my own p2sgrad implementation.  
NaN did not happen.
```
Using P2SHead
mintDCF: 0.065265       EER: 2.502 %  
```

```
Using p2sgrad in my git, ran it multiple times with different random seeds
mintDCF: 0.061463       EER: 2.462 % 
mintDCF: 0.056467       EER: 2.309 % 
mintDCF: 0.073829       EER: 3.087 % 
...
```

This is the training log. 
```
--------------------------------------------------------------             
  Epoch |  Duration(s) |   Train loss |     Dev loss |  Best               
--------------------------------------------------------------             
      0 |         53.7 |       0.0358 |       0.0049 |   yes               
      1 |         52.9 |       0.0033 |       0.0036 |   yes               
      2 |         52.6 |       0.0017 |       0.0023 |   yes               
      3 |         53.0 |       0.0022 |       0.0021 |   yes               
      4 |         53.0 |       0.0008 |       0.0026 |    no               
      5 |         53.0 |       0.0013 |       0.0044 |    no               
      6 |         53.0 |       0.0005 |       0.0016 |   yes               
      7 |         52.9 |       0.0014 |       0.0010 |   yes               
      8 |         53.0 |       0.0009 |       0.0031 |    no               
      9 |         53.0 |       0.0009 |       0.0013 |    no LR -> 1.50e-04
     10 |         52.8 |       0.0002 |       0.0010 |   yes LR -> 1.50e-04
     11 |         52.7 |       0.0001 |       0.0015 |    no LR -> 1.50e-04
     12 |         52.8 |       0.0001 |       0.0010 |    no LR -> 1.50e-04
     13 |         52.9 |       0.0001 |       0.0012 |    no LR -> 1.50e-04
     14 |         52.9 |       0.0002 |       0.0014 |    no LR -> 1.50e-04
     15 |         52.8 |       0.0001 |       0.0007 |   yes LR -> 1.50e-04
     16 |         52.8 |       0.0005 |       0.0025 |    no LR -> 1.50e-04
     17 |         52.8 |       0.0004 |       0.0014 |    no LR -> 1.50e-04
     18 |         52.7 |       0.0001 |       0.0012 |    no LR -> 1.50e-04
     19 |         52.8 |       0.0002 |       0.0011 |    no LR -> 7.50e-05
     20 |         53.0 |       0.0001 |       0.0008 |    no LR -> 7.50e-05
     21 |         53.1 |       0.0000 |       0.0007 |    no LR -> 7.50e-05
     22 |         52.9 |       0.0001 |       0.0009 |    no LR -> 7.50e-05
     23 |         52.9 |       0.0000 |       0.0018 |    no LR -> 7.50e-05
     24 |         52.9 |       0.0002 |       0.0008 |    no LR -> 7.50e-05
     25 |         53.0 |       0.0000 |       0.0011 |    no LR -> 7.50e-05
     26 |         52.9 |       0.0000 |       0.0009 |    no LR -> 7.50e-05
     27 |         52.8 |       0.0001 |       0.0010 |    no LR -> 7.50e-05
     28 |         52.9 |       0.0000 |       0.0008 |    no LR -> 7.50e-05
     29 |         52.9 |       0.0000 |       0.0009 |    no LR -> 3.75e-05
     30 |         52.8 |       0.0000 |       0.0007 |   yes LR -> 3.75e-05
     31 |         52.8 |       0.0000 |       0.0008 |    no LR -> 3.75e-05
     32 |         52.8 |       0.0000 |       0.0009 |    no LR -> 3.75e-05
     33 |         52.9 |       0.0000 |       0.0008 |    no LR -> 3.75e-05
     34 |         53.0 |       0.0000 |       0.0009 |    no LR -> 3.75e-05
     35 |         53.1 |       0.0000 |       0.0008 |    no LR -> 3.75e-05
     36 |         53.2 |       0.0000 |       0.0008 |    no LR -> 3.75e-05
     37 |         53.2 |       0.0000 |       0.0010 |    no LR -> 3.75e-05
     38 |         53.1 |       0.0000 |       0.0011 |    no LR -> 3.75e-05
     39 |         53.2 |       0.0000 |       0.0009 |    no LR -> 1.87e-05
     40 |         53.1 |       0.0000 |       0.0009 |    no LR -> 1.87e-05
     41 |         53.3 |       0.0000 |       0.0009 |    no LR -> 1.87e-05
...
     96 |         53.2 |       0.0000 |       0.0010 |    no LR -> 5.86e-07
--------------------------------------------------------------
Training finished by early stopping
Model is saved to./trained_network.pt
```

**Comment by TonyWangX** [comment]
Therefore, I think the P2SHead implementation is good.
The MSE loss on angles should not be a more difficult criteria than cos-face margin.

The optimization process may cause the issue. Could you check:
1. check whether self.weight in P2SHead is NaN or not;
2. check whether the input for forward() is NaN or not.

The P2SHead linear layer (self.weight) has no constraint / penalty applied. The output of the linear transformation can be any thing from -\infty to \infty.  This may be sensitive to the data you've used.

**Comment by LuckyZxy182** [comment]
Thanks for your reply and suggestions. I' ll check these carefully. Maybe the large classes_num causes Underfitting.
